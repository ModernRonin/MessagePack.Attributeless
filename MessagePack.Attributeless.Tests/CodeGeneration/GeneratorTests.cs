using ApprovalTests;
using FluentAssertions;
using MessagePack.Attributeless.CodeGeneration;
using NUnit.Framework;

namespace MessagePack.Attributeless.Tests.CodeGeneration;

[TestFixture]
public class GeneratorTests
{
    class BadType
    {
        public TestActionAttribute UnconfiguredType { get; set; }
    }

    [Test]
    public void Does_not_throw_for_valid_configuration()
    {
        var config = MessagePackSerializer.DefaultOptions.Configure()
            .GraphOf<Samples.IAnimal>()
            .Configuration;

        var underTest = new Generator();

        var action = () => underTest.Generate(config);

        action.Should().NotThrow();
    }

    [Test]
    public void Generate_concrete_type_with_enum()
    {
        var config = MessagePackSerializer.DefaultOptions.Configure()
            .GraphOf<Samples.Leg>()
            .Configuration;

        var result = new Generator().Generate(config);

        Approvals.Verify(result);
    }

    [Test]
    public void Generate_type_hierarchy()
    {
        var config = MessagePackSerializer.DefaultOptions.Configure()
            .GraphOf<Samples.IExtremity>()
            .Configuration;

        var result = new Generator().Generate(config);

        Approvals.Verify(result);
    }

    [Test]
    public void Throws_for_invalid_configuration()
    {
        var config = MessagePackSerializer.DefaultOptions.Configure()
            .GraphOf<BadType>()
            .Configuration;

        var underTest = new Generator();

        var action = () => underTest.Generate(config);

        action.Should()
            .Throw<ArgumentException>()
            .WithMessage(
                "configuration is incomplete - the following types are neither mapped nor handled by MessagePack natively nor enums: NUnit.Framework.TestActionAttribute");
    }
}