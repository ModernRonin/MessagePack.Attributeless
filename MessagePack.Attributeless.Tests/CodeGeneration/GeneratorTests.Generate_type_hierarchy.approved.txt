namespace Generated 
{
	using MessagePack.Formatters;

	public class SideFormatter: IMessagePackFormatter<MessagePack.Attributeless.Tests.Samples.Side>
	{
		public MessagePack.Attributeless.Tests.Samples.Side Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
		{
			return (MessagePack.Attributeless.Tests.Samples.Side) reader.ReadInt32((int) value)()
		}
		public void Serialize(ref MessagePackWriter writer, MessagePack.Attributeless.Tests.Samples.Side value, MessagePackSerializerOptions options) 
		{ 
			writer.WriteInt32(value);
		}
	}
}
namespace Generated 
{
	using MessagePack.Formatters;

	public class IExtremityFormatter: IMessagePackFormatter<MessagePack.Attributeless.Tests.Samples.IExtremity>
	{
		
		public MessagePack.Attributeless.Tests.Samples.IExtremity Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
		{
			if (reader.TryReadNil()) return default;

			var key = reader.ReadInt32();
			switch (key)
			{			
				
				case 0: return deserialize<MessagePack.Attributeless.Tests.Samples.Arm>();
				
				
				case 1: return deserialize<MessagePack.Attributeless.Tests.Samples.Leg>();
				
				
				case 2: return deserialize<MessagePack.Attributeless.Tests.Samples.Wing>();
				
				
			}
			throw new MessagePackSerializationException(
                    $"Encountered unknown type key {key} for MessagePack.Attributeless.Tests.Samples.IExtremity - was this serialized with a different configuration?");
			
			T deserialize<T>() where T: MessagePack.Attributeless.Tests.Samples.IExtremity 
            	=> return options.Resolver.GetFormatterWithVerify<T>().Deserialize(ref reader, options);
		}
		public void Serialize(ref MessagePackWriter writer, MessagePack.Attributeless.Tests.Samples.IExtremity value, MessagePackSerializerOptions options) 
		{ 
			if (value == null)
            {
                writer.WriteNil();
                return;
            }

		}
	}
}
